man/qwhich.1
.\" generated by mkman.py, please do NOT edit!
.TH qwhich "1" "Sep 2026" "Gentoo Foundation" "qwhich"
.SH NAME
qwhich \- find path to pkg
.SH SYNOPSIS
.B qwhich
\fI[opts] <pkg ...>\fR
.SH DESCRIPTION
.I qwhich
returns the paths to the ebuilds matching the given atoms.  Searching is
done by default over the main ebuild repository and all configured
overlays.  When requested, a search can be made over the repository of
installed packages (VDB) or binpkg repos.
.P
The default action is to print the full path to the latest ebuild for
each package.  This can be changed using \fB-d\fR to print the directory
the ebuild is in instead.  E.g\. one could \fBcd\fR(1) to this
directory.  Alternatively, one can request to stop searching after the
first match using \fB-f\fR, which means at most one match per repo is
returned.
.P
\fIqwhich\fR accepts atoms to search for, which can be any valid atom,
including e.g\. version specifications to narrow the search.
.P
The \fB-r\fR or \fB--repo\fR argument was removed in favour of the
global \fB--overlay\fR option.
.SH OPTIONS
.TP
\fB\-I\fR, \fB\-\-vdb\fR
Look in VDB (installed packages).
.TP
\fB\-b\fR, \fB\-\-binpkg\fR
Look at binary packages.
.TP
\fB\-t\fR, \fB\-\-tree\fR
Look in main tree and overlays.
.TP
\fB\-p\fR, \fB\-\-pretty\fR
Print (pretty) atom instead of path.

The default print format is \fI%[CATEGORY]%[PN]\fR or
\fI%[CATEGORY]%[PF]\fR in verbose mode (\fB-v\fR|\fB--verbose\fR).
The format can be changed with \fB-F\fR|\fB--format\fR.
.TP
\fB\-d\fR, \fB\-\-dir\fR
Print directory instead of path.
.TP
\fB\-R\fR, \fB\-\-printrepo\fR
Print repository name instead of path for tree/overlay matches.
.TP
\fB\-f\fR, \fB\-\-first\fR
Stop searching after first match (implies -l).
.TP
\fB\-l\fR, \fB\-\-latest\fR
Only return latest version for each match.
.TP
\fB\-T\fR, \fB\-\-novirtual\fR
Skip virtual category.
.TP
\fB\-A\fR, \fB\-\-noacct\fR
Skip acct-user and acct-group categories.
.TP
\fB\-F\fR \fI<arg>\fR, \fB\-\-format\fR \fI<arg>\fR
Print matched atom using given format string.

Setting this option has the effect of also setting
\fB-p\fR|\fB--pretty\fR.

Note that this only sets the format of the atom field, not the
entire output line.

For more information see \fBqatom\fR(1).
.TP
\fB\-\-root\fR \fI<arg>\fR
Set the ROOT env var.
.TP
\fB\-\-overlay\fR \fI<arg>\fR
Select the given overlay for use, instead of all found (see q -o).
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Report full package versions, emit more elaborate output.
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Tighter output; suppress warnings.
.TP
\fB\-C\fR, \fB\-\-nocolor\fR
Don't output color.
.TP
\fB\-\-color\fR
Force color in output.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print this help and exit.
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version and exit.

.SH "REPORTING BUGS"
Please report bugs via http://bugs.gentoo.org/
.br
Product: Gentoo Linux; Component: Current packages
.SH AUTHORS
.nf
Ned Ludd <solar@gentoo.org>
Mike Frysinger <vapier@gentoo.org>
Fabian Groffen <grobian@gentoo.org>
Jaeger Hofer <antiq.hofer@gmail.com>
.fi
.SH "SEE ALSO"



qwhich.c
/*
 * Copyright 2021-2026 Gentoo Foundation
 * Distributed under the terms of the GNU General Public License v2
 *
 * Copyright 2021-     Fabian Groffen  - <grobian@gentoo.org>
 * Copyright 2026-     Jaeger H.       - <antiq.hofer@gmail.com>
 */

#include "main.h"
#include "applets.h"

#include <string.h>
#include <time.h>
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <limits.h>
#include <sys/stat.h>

#include "atom.h"
#include "tree.h"

#define QWHICH_FLAGS "Ibtpdr:RflTAF:" COMMON_FLAGS
static struct option const qwhich_long_opts[] = {
	{"vdb",       no_argument, NULL, 'I'},
	{"binpkg",    no_argument, NULL, 'b'},
	{"tree",      no_argument, NULL, 't'},
	{"pretty",    no_argument, NULL, 'p'},
	{"dir",       no_argument, NULL, 'd'},
	{"printrepo", no_argument, NULL, 'R'},
	{"first",     no_argument, NULL, 'f'},
	{"latest",    no_argument, NULL, 'l'},
	{"novirtual", no_argument, NULL, 'T'},
	{"noacct",    no_argument, NULL, 'A'},
	{"format",     a_argument, NULL, 'F'},
	COMMON_LONG_OPTS
};
static const char * const qwhich_opts_help[] = {
	"Look in VDB (installed packages)",
	"Look at binary packages",
	"Look in main tree and overlays",
	"Print (pretty) atom instead of path for use with -F",
	"Print directory instead of path",
	"Print repository name instead of path for tree/overlay matches",
	"Stop searching after first match (implies -l)",
	"Only return latest version for each match",
	"Skip virtual category",
	"Skip acct-user and acct-group categories",
	"Print matched using given format string",
	COMMON_OPTS_HELP
};
static const char qwhich_desc[] = "Find paths to ebuilds.";
#define qwhich_usage(ret) \
	usage(ret, QWHICH_FLAGS, qwhich_long_opts, qwhich_opts_help, qwhich_desc, lookup_applet_idx("qwhich"))

struct qwhich_mode {
	bool do_vdb:1;
	bool do_binpkg:1;
	bool do_tree:1;
	bool print_atom:1;
	bool print_path:1;
	bool print_repo:1;
	bool match_first:1;
	bool match_latest:1;
	bool skip_virtual:1;
	bool skip_acct:1;
	const char *fmt;
};

int qwhich_main(int argc, char **argv)
{
	depend_atom *atom;
	array *atoms = array_new();
	array *trees = array_new();
	struct qwhich_mode m;
	array *tmc;
	tree_pkg_ctx *tmcw;
	size_t i;
	size_t j;
	char *overlay;
	size_t n;
	int ret;
	tree_ctx *t;
	char *reponam;
	size_t repolen;
	size_t pathlen;

	VAL_CLEAR(m);

	while ((ret = GETOPT_LONG(QWHICH, qwhich, "")) != -1) {
		switch (ret) {
			COMMON_GETOPTS_CASES(qwhich)

			case 'I': m.do_vdb = true;       break;
			case 'b': m.do_binpkg = true;    break;
			case 't': m.do_tree = true;      break;
			case 'p': m.print_atom = true;   break;
			case 'd': m.print_path = true;   break;
			case 'R': m.print_repo = true;   break;
			case 'f': m.match_first = true;  break;
			case 'l': m.match_latest = true; break;
			case 'T': m.skip_virtual = true; break;
			case 'A': m.skip_acct = true;    break;
			case 'F': m.fmt = optarg;        break;
		}
	}

	/* defaults: no selectors, enable tree + overlays */
	if (!m.do_vdb && !m.do_binpkg && !m.do_tree)
		m.do_tree = true;

	/* when printing path, we better just match the first, else we get a
	 * lot of dups */
	if (m.print_path || m.match_first)
		m.match_latest = true;

	/* set format if none given */
	if (m.fmt == NULL) {
		if (verbose)
			m.fmt = "%[CATEGORY]%[PF]";
		else
			m.fmt = "%[CATEGORY]%[PN]";
		/* pointless to report duplicates */
		if (m.print_atom &&
			!verbose)
			m.match_latest = true;
	} else {
		/* makes no sense to use formatter if we're not going to use it */
		m.print_atom = true;
	}

	argc -= optind;
	argv += optind;
	for (i = 0; i < (size_t)argc; ++i) {
		atom = atom_explode(argv[i]);
		if (!atom)
			warn("invalid atom: %s", argv[i]);
		else
			array_append(atoms, atom);
	}

	/* TODO: silence when the path doesn't exist -- reasonable though? */
	if (m.do_vdb) {
		t = tree_new(portroot, portvdb, TREETYPE_VDB, false);
		if (t != NULL)
			array_append(trees, t);
	}
	if (m.do_binpkg) {
		t = tree_new(portroot, pkgdir, TREETYPE_BINPKG, false);
		if (t != NULL)
			array_append(trees, t);
	}

	if (m.do_tree) {
		array_for_each(overlays, n, overlay) {
			t = tree_new(portroot, overlay, TREETYPE_EBUILD, false);
			if (t != NULL)
				array_append(trees, t);
		}
	}

	/* at least keep the IO constrained to a tree at a time */
	array_for_each(trees, j, t) {
		char *reponame = tree_get_repo_name(t);

		pathlen = strlen(tree_get_path(t));
		if (m.print_repo && reponame != NULL) {
			reponam = reponame;
			repolen = strlen(reponam);
		} else {
			reponam = tree_get_path(t);
			repolen = pathlen;
		}

		array_for_each(atoms, i, atom) {
			/* because this is for direct visuals, always require sorted
			 * order so the output is consistent and predictable */
			tmc = tree_match_atom(t, atom, TREE_MATCH_SORT |
					(m.match_latest ? TREE_MATCH_LATEST : 0 ) |
					(m.match_first  ? TREE_MATCH_FIRST  : 0 ) |
					(m.skip_virtual ? 0 : TREE_MATCH_VIRTUAL) |
					(m.skip_acct    ? 0 : TREE_MATCH_ACCT   ));
			array_for_each(tmc, n, tmcw)
			{
				if (m.print_atom) {
					printf("%s\n", atom_format(m.fmt,
											   tree_pkg_atom(tmcw, false)));
				} else {
					char  *path = tree_pkg_get_path(tmcw) + 1;
					size_t len  = strlen(path);

					if (len > pathlen)
					{
						path += pathlen;
						len  -= pathlen;
					}

					if (m.print_path)
					{
						char *p = strrchr(path, '/');
						if (p == NULL)
							len = 0;
						else
							len = p - path;
					}

					printf("%s%s%.*s%s%s%.*s%s\n",
						   GREEN, m.print_repo ? "" : "/",
						   (int)repolen, reponam, m.print_repo ? "::" : "/",
						   DKBLUE, (int)len, path, NORM);
				}
			}
			array_free(tmc);
		}
		tree_close(t);
	}
	array_deepfree(atoms, atom_implode_cb);
	array_free(trees);

	return EXIT_SUCCESS;
}
